src/AzureBlobStorageUpdateSession/UpdateSession/UpdateSessionTable.cs

```diff
@@ -0,0 +1,123 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Table;
+
+namespace Etg.Yams.Azure.UpdateSession
+{
+    public class UpdateSessionTable : IUpdateSessionTable
+    {
+        private readonly TimeSpan _ttl;
+        private readonly CloudTable _table;
+        public const string UpdateSessionTableName = "YamsUpdateSession";
+        public const string UpdateDomainRowKey = "updateDomain";
+        public const string ModifiedRowKey = "modified";
+
+        /// <summary>
+        /// </summary>
+        /// <param name="connectionString"></param>
+        /// <param name="ttl">If an instance update session is older than ttl, it will be ignored. The goal of this parameter 
+        /// is to avoid situations where an instance fails to end its update session and blocks the cluster.</param>
+        public UpdateSessionTable(string connectionString, TimeSpan ttl)
+        {
+            _ttl = ttl;
+            CloudTableClient client = CloudStorageAccount.Parse(connectionString).CreateCloudTableClient();
+            _table = client.GetTableReference(UpdateSessionTableName);
+            _table.CreateIfNotExists();
+        }
+
+        public Task<UpdateSessionStatus> FetchUpdateSessionStatus(string clusterId, string appId)
+        {
+            string partitionQuery = CreatePartitionQuery(clusterId, appId);
+
+            TableQuery<UpdateDomainEntity> query = new TableQuery<UpdateDomainEntity>().Where(partitionQuery);
+            UpdateDomainEntity updateDomainEntity = null;
+            UpdateDomainEntity instanceListModifiedEntity = null;
+            List<UpdateDomainEntity> instancesList = new List<UpdateDomainEntity>();
+            foreach (UpdateDomainEntity entity in _table.ExecuteQuery(query))
+            {
+                if (entity.RowKey == UpdateDomainRowKey)
+                {
+                    updateDomainEntity = entity;
+                }
+                else if (entity.RowKey == ModifiedRowKey)
+                {
+                    instanceListModifiedEntity = entity;
+                }
+                else
+                {
+                    if (entity.Timestamp.Add(_ttl) > DateTimeOffset.Now)
+                    {
+                        instancesList.Add(entity);
+                    }
+                }
+            }
+
+            IEnumerable<UpdateDomainEntity> instancesEntities = instancesList;
+            if (updateDomainEntity != null)
+            {
+                // filter out entities that do not match the active update domain (were leftover for some reason)
+                instancesEntities = instancesEntities.Where(entity => entity.UpdateDomain == updateDomainEntity.UpdateDomain);
```

Comment:
Avoid iterating twice maybe by adding the check in the foreach above?

Reply:
Not possible because the updateDomainEntity must be found first.

Reply:
Can't you run a query to find the entity with the RowKey equal to UpdateDomainRowKey and then another query maybe to find the non-expired ones matching the UpdateDomain? Not a big concern but if you're querying over indexed facets it will likely be faster than iterating for big clusters. Your call.